[WebRTC] Add support for legacy Offer options

philn opened the pull request:
#### ef7ef86e4cf56ad91d0a06677ba8de5de8cc3f1e
<pre>
[WebRTC] Add support for legacy Offer options
<a href="https://bugs.webkit.org/show_bug.cgi?id=256707">https://bugs.webkit.org/show_bug.cgi?id=256707</a>
&lt;<a href="https://rdar.apple.com/problem/109569164">rdar://problem/109569164</a>&gt;

Reviewed by Youenn Fablet.

Some websites still rely on the legacy offer options, despite the spec recommendation of using
transceivers, so bring back those options, behind a runtime WebPreference, disabled by default.

Covered by new tests backported from WPT.
Upstream commit: <a href="https://github.com/web-platform-tests/wpt/commit/9c29d4eefd4d5d41822f6fdccaf59b14f0603965">https://github.com/web-platform-tests/wpt/commit/9c29d4eefd4d5d41822f6fdccaf59b14f0603965</a>

* LayoutTests/TestExpectations:
* LayoutTests/platform/glib/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
(WebCore::PeerConnectionBackend::addTransceiver):
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.h:
* Source/WebCore/Modules/mediastream/RTCOfferOptions.h:
* Source/WebCore/Modules/mediastream/RTCOfferOptions.idl:
* Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
(WebCore::RTCPeerConnection::addReceiveOnlyTransceiver):
(WebCore::RTCPeerConnection::createOffer):
* Source/WebCore/Modules/mediastream/RTCPeerConnection.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::createTransceiverBackends):
(WebCore::GStreamerMediaEndpoint::addTransceiver):
(WebCore::GStreamerMediaEndpoint::onNegotiationNeeded):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
(WebCore::GStreamerPeerConnectionBackend::addTransceiverFromTrackOrKind):
(WebCore::GStreamerPeerConnectionBackend::addTransceiver):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
(WebCore::LibWebRTCMediaEndpoint::createTransceiverBackends):
(WebCore::LibWebRTCMediaEndpoint::addTransceiver):
(WebCore::LibWebRTCMediaEndpoint::OnNegotiationNeededEvent):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCPeerConnectionBackend.cpp:
(WebCore::LibWebRTCPeerConnectionBackend::addTransceiverFromTrackOrKind):
(WebCore::LibWebRTCPeerConnectionBackend::addTransceiver):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCPeerConnectionBackend.h:

Canonical link: <a href="https://commits.webkit.org/283089@main">https://commits.webkit.org/283089@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d1bcaeceab6503f719b1f122fc3182d5d0f11ffe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65189 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44556 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17803 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69213 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15795 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67307 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52339 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16077 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52373 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10931 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68255 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41189 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56429 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32998 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37860 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13800 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14671 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/58302 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14139 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70918 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/64432 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9141 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13619 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59702 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9173 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56489 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59976 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14374 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7556 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1228 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/86200 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40368 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15185 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41445 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42626 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41189 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/Modules/mediastream/RTCOfferOptions.idl:
```diff
@@ -28,4 +28,7 @@
 ] dictionary RTCOfferOptions : RTCOfferAnswerOptions {
     boolean iceRestart = false;
+
+    [EnabledBySetting=LegacyWebRTCOfferOptionsEnabled] boolean? offerToReceiveVideo;
+    [EnabledBySetting=LegacyWebRTCOfferOptionsEnabled] boolean? offerToReceiveAudio;
```
If there are too many concerns about this maybe these can be exposed only for GLib ports....

youennf commented on Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
```diff
@@ -215,6 +215,42 @@ void RTCPeerConnection::createOffer(RTCOfferOptions&& options, Ref<DeferredPromi
+    };
+
+    if (options.offerToReceiveVideo)
+        processLegacyOption(options.offerToReceiveVideo.value(), "video"_s);
```
```suggestion
        processLegacyOption(*options.offerToReceiveVideo, "video"_s);
```

youennf commented on Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
```diff
@@ -215,6 +215,42 @@ void RTCPeerConnection::createOffer(RTCOfferOptions&& options, Ref<DeferredPromi
+        processLegacyOption(options.offerToReceiveVideo.value(), "video"_s);
+
+    if (options.offerToReceiveAudio)
+        processLegacyOption(options.offerToReceiveAudio.value(), "audio"_s);
```
```suggestion
        processLegacyOption(*options.offerToReceiveAudio, "audio"_s);
```

youennf commented on Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
```diff
@@ -215,6 +215,42 @@ void RTCPeerConnection::createOffer(RTCOfferOptions&& options, Ref<DeferredPromi
+        }
+
+        RTCRtpTransceiverInit init { .direction = RTCRtpTransceiverDirection::Recvonly, .streams = { }, .sendEncodings = { } };
+        addTransceiver(trackKind, init);
```
This will likely triggers negotiation needed flag. It might be best to pipe it platform specific code that will handle this part properly.

philn replied:
Hm yes, this needs some work indeed. I noticed a couple more issues. There's also a couple WPT tests related to these options too. I'll rework this PR. Thanks for the review!

youennf commented on Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
```diff
@@ -215,6 +215,42 @@ void RTCPeerConnection::createOffer(RTCOfferOptions&& options, Ref<DeferredPromi
         return;
     }
 
+    // https://www.w3.org/TR/webrtc/#legacy-configuration-extensions
```
There are likely WPT tests that would benefit of this.
Can you verify this?

philn replied:
I only found `imported/w3c/web-platform-tests/webrtc/RTCPeerConnection-ontrack.https.html` but it's already passing with current main... 🤔

youennf reviewed (commented):
Some comments.
Overall looks good with a flag.
It would be good to have tests and check how to handle negotiation needed.

youennf commented on Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
```diff
@@ -207,6 +207,21 @@ ExceptionOr<Ref<RTCRtpTransceiver>> RTCPeerConnection::addTransceiver(AddTransce
+    if (kind != "audio"_s && kind != "video"_s)
+        return Exception { ExceptionCode::TypeError };
+
+    if (isClosed())
```
Maybe ASSERT(!isClosed()) ?

youennf commented on Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCPeerConnectionBackend.h:
```diff
@@ -95,7 +95,7 @@ class LibWebRTCPeerConnectionBackend final : public PeerConnectionBackend {
     void removeTrack(RTCRtpSender&) final;
 
-    ExceptionOr<Ref<RTCRtpTransceiver>> addTransceiver(const String&, const RTCRtpTransceiverInit&) final;
+    ExceptionOr<Ref<RTCRtpTransceiver>> addTransceiver(const String&, const RTCRtpTransceiverInit&, IgnoreNegotiationNeededFlag = IgnoreNegotiationNeededFlag::No) final;
```
Do we need the ` = IgnoreNegotiationNeededFlag::No`?

philn replied:
It can be made explicit, no problem :)

youennf commented on Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCPeerConnectionBackend.cpp:
```diff
@@ -331,9 +331,9 @@ ExceptionOr<Ref<RTCRtpSender>> LibWebRTCPeerConnectionBackend::addTrack(MediaStr
 
 template<typename T>
-ExceptionOr<Ref<RTCRtpTransceiver>> LibWebRTCPeerConnectionBackend::addTransceiverFromTrackOrKind(T&& trackOrKind, const RTCRtpTransceiverInit& init)
+ExceptionOr<Ref<RTCRtpTransceiver>> LibWebRTCPeerConnectionBackend::addTransceiverFromTrackOrKind(T&& trackOrKind, const RTCRtpTransceiverInit& init, IgnoreNegotiationNeededFlag ignoreNegotiationNeededFlag)
```
Let's instead ASSERT that ignoreNegotiationNeededFlag is `No`.
I do not think this works for libwebrtc since the negotiation needed flag might be triggered asynchronously, for instance after the chaining of operations.

philn replied:
> I do not think this works for libwebrtc since the negotiation needed flag might be triggered asynchronously, for instance after the chaining of operations.

I have verified that `LibWebRTCMediaEndpoint::OnNegotiationNeededEvent()` returns early when legacy offer options are processed, using the layout test `http/wpt/webrtc/RTCPeerConnection-createOffer-offerToReceive.html`.

I also tried a simpler approach, by setting the libwebrtc legacy offer options directly in `LibWebRTCMediaEndpoint::doCreateOffer()` without prior management in the top-level RTCPeerConnection, but that triggered various errors in the layout test.

I think the current approach of this PR works as expected :)

youennf reviewed (approved):
OK, let's import WPT webrtc/legacy then and verify they pass as expected with this patch (ideally with libwebrtc backend too).

philn commented on LayoutTests/imported/w3c/web-platform-tests/webrtc/legacy/w3c-import.log:
```diff
@@ -0,0 +1,22 @@
+The tests in this directory were imported from the W3C repository.
+Do NOT modify these tests directly in WebKit.
+Instead, create a pull request on the WPT github:
+	https://github.com/web-platform-tests/wpt
```
Oh no, does this mean the `<!-- webkit-test-runner [ LegacyWebRTCOfferOptionsEnabled=true ] -->` preambles manually added will disappear at next import?

philn replied:
The answer is no, thanks Ms2ger.

philn committed:
[WebRTC] Add support for legacy Offer options
https://bugs.webkit.org/show_bug.cgi?id=256707
<rdar://problem/109569164>

Reviewed by Youenn Fablet.

Some websites still rely on the legacy offer options, despite the spec recommendation of using
transceivers, so bring back those options, behind a runtime WebPreference, disabled by default.

Covered by new tests backported from WPT.
Upstream commit: https://github.com/web-platform-tests/wpt/commit/9c29d4eefd4d5d41822f6fdccaf59b14f0603965

* LayoutTests/TestExpectations:
* LayoutTests/platform/glib/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
(WebCore::PeerConnectionBackend::addTransceiver):
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.h:
* Source/WebCore/Modules/mediastream/RTCOfferOptions.h:
* Source/WebCore/Modules/mediastream/RTCOfferOptions.idl:
* Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
(WebCore::RTCPeerConnection::addReceiveOnlyTransceiver):
(WebCore::RTCPeerConnection::createOffer):
* Source/WebCore/Modules/mediastream/RTCPeerConnection.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::createTransceiverBackends):
(WebCore::GStreamerMediaEndpoint::addTransceiver):
(WebCore::GStreamerMediaEndpoint::onNegotiationNeeded):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
(WebCore::GStreamerPeerConnectionBackend::addTransceiverFromTrackOrKind):
(WebCore::GStreamerPeerConnectionBackend::addTransceiver):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
(WebCore::LibWebRTCMediaEndpoint::createTransceiverBackends):
(WebCore::LibWebRTCMediaEndpoint::addTransceiver):
(WebCore::LibWebRTCMediaEndpoint::OnNegotiationNeededEvent):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCPeerConnectionBackend.cpp:
(WebCore::LibWebRTCPeerConnectionBackend::addTransceiverFromTrackOrKind):
(WebCore::LibWebRTCPeerConnectionBackend::addTransceiver):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCPeerConnectionBackend.h:

Canonical link: https://commits.webkit.org/283089@main
